fdbserver/ClusterController.actor.cpp

```diff
@@ -2476,9 +2476,63 @@ ACTOR Future<Void> monitorClientTxnInfoConfigs(ClusterControllerData::DBInfo* db
 			try {
 				tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
 				tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+				state Optional<Value> globalConfigVersion = wait(tr.get(globalConfigVersionKey));
 				state Optional<Value> rateVal = wait(tr.get(fdbClientInfoTxnSampleRate));
 				state Optional<Value> limitVal = wait(tr.get(fdbClientInfoTxnSizeLimit));
-				ClientDBInfo clientInfo = db->clientInfo->get();
+				state ClientDBInfo clientInfo = db->clientInfo->get();
+
+				if (globalConfigVersion.present()) {
+					BinaryReader versionReader = BinaryReader(globalConfigVersion.get(), AssumeVersion(g_network->protocolVersion()));
+					int64_t commitVersion;
+					int16_t serializationOrder;
+					state ProtocolVersion protocolVersion;
+					versionReader >> commitVersion >> serializationOrder >> protocolVersion;
+
+					state Arena arena;
+					if (protocolVersion == g_network->protocolVersion()) {
+						Standalone<RangeResultRef> globalConfigHistory = wait(tr.getRange(globalConfigHistoryKeys, CLIENT_KNOBS->TOO_MANY));
+						// If the global configuration version key has been
+						// set, the history should contain at least one item.
+						ASSERT(globalConfigHistory.size() > 0);
+						clientInfo.history.clear();
+
+						for (const auto& kv : globalConfigHistory) {
+							Standalone<std::pair<Version, VectorRef<MutationRef>>> data;
+
+							// Read commit version out of versionstamp at end of key.
+							BinaryReader versionReader = BinaryReader(kv.key.removePrefix(globalConfigHistoryPrefix), AssumeVersion(protocolVersion));
+							Version historyCommitVersion;
+							versionReader >> historyCommitVersion;
+							historyCommitVersion = bigEndian64(historyCommitVersion);
+							data.first = historyCommitVersion;
+
+							// Read the list of mutations that occurred at this version.
+							BinaryReader mutationReader = BinaryReader(kv.value, AssumeVersion(protocolVersion));
+							VectorRef<MutationRef> mutations;
+							mutationReader >> mutations;
+							data.second = VectorRef(arena, mutations);
+
+							clientInfo.history.push_back(data);
+						}
+
+						// History should be ordered by version, ascending.
+						std::sort(clientInfo.history.begin(), clientInfo.history.end(), [](const auto& lhs, const auto& rhs) {
+							return lhs.first < rhs.first;
+						});
+					} else {
+						// If the protocol version has changed, the
+						// GlobalConfig actor should refresh its view by
+						// reading the entire global configuration key range.
+						// An empty mutation list will signal the actor to
+						// refresh.
+						clientInfo.history.clear();
+					}
+
+					clientInfo.id = deterministicRandom()->randomUniqueID();
+					db->clientInfo->set(clientInfo);
+				}
+
+				// TODO: Remove this and move to global config space
```

Comment:
A future PR will address this.

Reply:
This PR now includes these changes (starting with https://github.com/apple/foundationdb/pull/4330/commits/38fdfbaba3456b9c408e6669fd34d169056fd8ec).